### ISSUE TITLE
Support for Tiny C Compiler

### ISSUE THREAD

**Issue by Tilka** [compiler-explorer/compiler-explorer#246]
Compiler Explorer should obviously support the smolest and therefore without question best C compiler in the world: the [Tiny C Compiler](http://bellard.org/tcc/).

**Comment by haneefmubarak** [comment]
Since TCC was already listed on the master list, I decided to have a look at what it would take to implement this. A bit of searching revealed that TCC [directly generates binary opcodes without using an intermediate assembler](https://lists.nongnu.org/archive/html/tinycc-devel/2006-11/msg00008.html):

>> Hi,
>>         I want to implement a register allocation
>> algorithm.
>
>TinyCC is not prepared for sophisticated register allocation algorithms.
>Registers are written back to memory after every statement.
>
>> Please let me know 
>> 1. Which module/file deals with register allocation.
>
>tcc.c: from function save_reg() down to function gv2()
>
>> 2. How to generate assembly program in tcc.
>
>You don't.
>The i386 and ARM targets use a function o() to write the opcodes as they
>appear in the binary. Use objdump -d if you need assembly output.

Indeed, checking [the options sections of the TCC documentation](http://bellard.org/tcc/tcc-doc.html#SEC4) confirms this: there aren't any options for outputting assembly. Even if we were to go ahead and just disassemble the output from TCC, we wouldn't be able to obtain the main benefit of `compiler-explorer`, that is, the ability to see how code compiles to assembly on a line by line basis.

---

However, there is a potential, albeit involved, approach that could allow us to get TCC to work with `compiler-explorer` to actually get the line by line data that we want. Seeing as TCC essentially compiles things down on a statement-by-statement level (which, IMHO, could actually be really neat to use when teaching absolute beginners the basics of assembly data and control flow), we could patch the output stages of TCC's codegen to spit out assembly instead of opcodes, along with the line number of the statement that it's outputting instructions for.

This will involve a lot of work, if anyone wants to do it (unlikely), but on the positive side of things, this will be relatively straightforward to maintain due to the ~~~absolutely glacial crawl at which TCC development moves at present (and over the past few years)~~~ [relatively slow](http://repo.or.cz/git-browser/by-commit.html?r=tinycc.git) development and also the possibility of merging such changes back upstream into the TCC project.

In short, there are a few changes needed to TCC itself before decent `compiler-explorer` support can be implemented for it; whether that is or isn't in the scope of the `compiler-explorer` project is likely something that has yet to be decided, although I expect that the amount of work severely outweighs the benefit that would be yielded from gaining TCC support.

@mattgodbolt thoughts?

**Comment by mattgodbolt** [comment]
Hi @haneefmubarak - thanks so much for taking the time to investigate! Indeed, this is why compilers with binary output are harder to implement in CE.

That said; since I first looked at `tcc`, we've put in binary support. That is, for some compilers I compile to final output, and then disassemble to get the results. If we can do this with `tcc` then we're home and dry.

The big deal with the disassembly is that in order to get the line colourisation working one needs the debug output too: hopefully `tcc` does use DWARF or something similar to let us map assembly back to the source.

**Comment by mattgodbolt** [comment]
Getting tcc running at all proved to be pretty simple (I spent 15 mins hacking on the train):

![tcc](https://cloud.githubusercontent.com/assets/633973/22332510/ed1d9b92-e396-11e6-9956-302283c6d54e.png)

Would need a bit more work to turn it into a real supported compiler. Specifically:

- support per-compiler "output filename" (I hacked it for now, but tcc understandably doesn't compile "foo.cpp")
- come up with a solution for the line stuff. `tcc` supports `-g` but the objdump disassembly doesn't have per-line stuff in it.
- probably make a new `compilerType` to handle always compiling in binary mode. At the moment if you deselect the binary button it will fail.

But...it's doable! :)

**Comment by haneefmubarak** [comment]
Ah! Totally forgot about debugging info, although if it's not spitting out line level stuff, I can't think of anything that we can do other than modifying TCC itself.

While at it, it might also be worthwhile to add a button that switches between compiling for C/C++ (that we can grey-out for things like TCC), because compilers like TCC don't support C++ and people might get a touch confused if the default is C++ and then one of the compilers spits out garbage errors when feeding in the same C++ code.

**Comment by mattgodbolt** [comment]
It turns out tcc does debug info! Yay! Sadly the debug info seems to get
lost when linking boo. But I have a plan for it :-)

Will keep it on the back burner

**Comment by jxu** [comment]
Do you still have any plans for this?

**Comment by mattgodbolt** [comment]
We could now probably support this (minus the colourisation/debug stuff) as we have a separate "C" language, which alleviates some of the problems seen. Happy for someone else to take a look though, I don't have too much time. https://github.com/mattgodbolt/compiler-explorer/blob/master/docs/AddingACompiler.md has some info!

**Comment by cnlohr** [comment]
I mean it sounds like you were _really_ close as it was.  Even if TCC has subpar support, it's a heck of a lot better than asking people to compile and then use objdump on their own when they're checking code optimization in TCC.

**Comment by partouf** [comment]
I'll put this on the calendar to do this or next week

**Comment by cnlohr** [comment]
@partouf This would be a huge boon. Also, I don't mind putting $20 toward this.

Also, some folks and I may be adding RV32 to TCC, and it would be important for us to have that variant then, it's easier to add another variant of a compiler than it is to add a brand new oddball compiler, like TCC correct?

**Comment by cnlohr** [comment]
Uuhhh ... is that really all it takes to add minimal support?  I don't understand where the link to the executable is actually coming from - like how those changes get the tinycc executable?

Re: different arches: Normally, you compile them as different executables, in fact, in a recent thread to the list they even recommended making 32-bit RISC-V as a separate executable from 64-bit RISC-V and they're very similar arches.

**Comment by partouf** [comment]
Yes, those are the changes that are needed for CE to support it. It might have been harder in the past, but all it just takes is someone to take a bit of time to make it work. (it's documented how to here: https://github.com/compiler-explorer/compiler-explorer/blob/master/docs/AddingACompiler.md)

What I had to do extra is create a buildscript and install it:
https://github.com/compiler-explorer/misc-builder/blob/main/build/build-tinycc.sh#L50

If you have any suggestions on how to add the other architectures, please do. Also feel free to send a PR if that's easier.

**Comment by partouf** [comment]
This is now live for x86-64

**Comment by cnlohr** [comment]
I will keep this thread bookmarked for when this comes through.

**Comment by cnlohr** [comment]
Hmm, I don't see it in the current list is that a regression?

**Comment by partouf** [comment]
you can search for it in the compiler list by typing tcc

![image](https://user-images.githubusercontent.com/442630/142068444-e133a957-5121-4af3-9d21-458583c697be.png)

**Comment by cnlohr** [comment]
![image](https://user-images.githubusercontent.com/2748168/142068555-f4e5d29f-9165-4399-afdc-a4e4f9faaf9b.png)

**Comment by partouf** [comment]
it's not in the C++ list, that's correct

**Comment by cnlohr** [comment]
Oh! I didn't notice that they are fundamentally different lists now.  That must have been my confusion for a long time.
